mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/task/SolveMessageInconsistenciesService.java

```diff
@@ -0,0 +1,92 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.cassandra.mail.task;
+
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.james.mailbox.cassandra.ids.CassandraId;
+import org.apache.james.mailbox.cassandra.ids.CassandraMessageId;
+import org.apache.james.mailbox.cassandra.mail.CassandraMessageIdDAO;
+import org.apache.james.mailbox.cassandra.mail.CassandraMessageIdToImapUidDAO;
+import org.apache.james.mailbox.model.ComposedMessageIdWithMetaData;
+import org.apache.james.task.Task;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class SolveMessageInconsistenciesService {
+
+    public static final Logger LOGGER = LoggerFactory.getLogger(SolveMessageInconsistenciesService.class);
+
+    private final CassandraMessageIdToImapUidDAO idToImapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+
+    @Inject
+    SolveMessageInconsistenciesService(CassandraMessageIdToImapUidDAO idToImapUidDAO, CassandraMessageIdDAO messageIdDAO) {
+        this.idToImapUidDAO = idToImapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+    }
+
+    Mono<Task.Result> fixMessageInconsistencies() {
+        //TODO: iterate MessageIdTable to remove redundant messages;
+        return Flux.merge(
+            fixMessageIdInconsistencies())
+            .reduce(Task.Result.COMPLETED, Task::combine);
+    }
+
+    private Mono<Task.Result> fixMessageIdInconsistencies() {
+        return idToImapUidDAO.retrieveAllMessages()
+            .flatMap(message -> idToImapUidDAO.retrieve((CassandraMessageId) message.getComposedMessageId().getMessageId(), Optional.of((CassandraId) message.getComposedMessageId().getMailboxId())))
+            .flatMap(this::fetchAndFixMessageId)
+            .reduce(Task.Result.COMPLETED, Task::combine);
+    }
+
+    private Mono<Task.Result> fetchAndFixMessageId(ComposedMessageIdWithMetaData message) {
+        return messageIdDAO.retrieve((CassandraId) message.getComposedMessageId().getMailboxId(), message.getComposedMessageId().getUid())
+            .filter(Optional::isPresent)
+            .flatMap(optional -> fixWhenMessageFoundInMessageId(message, optional.get()))
+            .switchIfEmpty(fixWhenMessageNotFoundInMessageId(message));
+    }
+
+    private Mono<Task.Result> fixWhenMessageFoundInMessageId(ComposedMessageIdWithMetaData messageFromImapUid, ComposedMessageIdWithMetaData messageFromMessageId) {
+        return Mono.fromCallable(() -> messageFromImapUid.equals(messageFromMessageId))
+            .filter(Boolean::booleanValue)
```

Comment:
Do a if

Reply:
`.flatMap(isEqual -> {
                if (isEqual) {
                    return Mono.just(Task.Result.COMPLETED);
                }

                return messageIdDAO.updateMetadata(messageFromImapUid)
                    .then(Mono.just(Task.Result.COMPLETED))
                    .onErrorResume(error -> {
                        LOGGER.error("Error when fixing inconsistency for message: {}", messageFromImapUid, error);
                        return Mono.just(Task.Result.PARTIAL);
                    });`

Looks ugly in my opinion :(